platform/linux-generic/include/odp_chksum_internal.h

```diff
@@ -0,0 +1,71 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP checksum - implementation internal
+ */
+
+#ifndef ODP_CHKSUM_INTERNAL_H_
+#define ODP_CHKSUM_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+
+/* Simple implementation of ones complement sum.
+ * Based on RFC1071 and its errata.
+ */
+typedef union {
+	uint16_t w;
+	uint8_t b[2];
+} swap_buf_t;
+
+static uint32_t _odp_chksum_ones_comp16_32(const void *p, uint32_t len,
+					   odp_bool_t odd_offset)
+
+{
+	uint32_t sum;
+	const uint16_t *data;
+
+	if (len > 0 && odd_offset) {
+		swap_buf_t sw;
+
+		sw.b[0] = 0;
+		sw.b[1] = *(const uint8_t *)p;
+		sum = sw.w;
+		data = (const uint16_t *)((uintptr_t)p+1);
+		len--;
+	} else {
+		sum = 0;
+		data = (const uint16_t *)p;
+	}
+
+	while (len > 1) {
+		sum += *data++;
+		len -= 2;
```

Comment:
This routine will still fail if `odd_offset` is true and `p` is even at entry. This will occur if a packet consists of two segments, the first of which has an odd number of bytes to be checksummed since ODP will start each segment on an even boundary. If will similarly fail if `odd_offset` is false and `p` is odd at entry. That would occur if the second segment had been truncated . 

`odd_offset` and the even/odd values of `p` are independent and need to be handled separately.

Reply:
Yes, I understand that. I'd like to point to `helper/chksum.c` and current `platform/l-g/odp_chksum.c`.

Reply:
This can be made robust without any real loss of clarity or performance:
```
static inline uint16_t data16_unaligned(void *p)
{
	return (*(uint8_t *)p << 8) | (*(uint8_t *)(p + 1));
}

static uint32_t _odp_chksum_ones_comp16_32(const void *p, uint32_t len,
					   odp_bool_t odd_offset)

{
	uint32_t sum;
	const uint16_t *data;

	if (odp_unlikely(len > 0 && odd_offset)) {
		swap_buf_t sw;

		sw.b[0] = 0;
		sw.b[1] = *(const uint8_t *)p;
		sum = sw.w;
		data = (const uint16_t *)((uintptr_t)p + 1);
		len--;
	} else {
		sum = 0;
		data = (const uint16_t *)p;
	}

	if (odp_unlikely(data % 2)) {
		while (len > 1) {
			sum += data16_unaligned(data++);
			len -= 2;
		}
	} else {
		while (len > 1) {
			sum += *data++;
			len -= 2;
		}
	}

	/* Add left-over byte, if any */
	if (len > 0) {
		swap_buf_t sw;

		sw.b[0] = *(const uint8_t *)data;
		sw.b[1] = 0;
		sum += sw.w;
	}

	return sum;
}
```